ユーザー管理機能　実装

nana0817 opened the pull request:
# What
Furimaアプリのユーザー管理機能を実装
# Why
Furimaアプリでユーザーを管理できるようにするため

 ### 必要な情報を適切に入力して「会員登録」ボタンを押すと、ユーザーの新規登録ができる動画
https://gyazo.com/0febaed81df980e201881d8fb083c2cf
 ### 入力に問題のある状態で「会員登録」ボタンが押された場合、ユーザーの新規登録はできず、エラーメッセージが出力される動画
https://gyazo.com/cf5204c0df0592751b9e34932c7acf5d
### 必要な情報を適切に入力して「ログイン」ボタンを押すと、ユーザーのログインができる動画
https://gyazo.com/015693e4887c6a9c96eef3bbd75dba5f
### 入力に問題のある状態で「ログイン」ボタンが押された場合、ユーザーのログインはできず、エラーメッセージが出力される動画
https://gyazo.com/af5b321f1f3e9ec9603f4efa0da06962
### ログアウトに成功する動画
https://gyazo.com/99bf97bd023bfc32c1cddd8603343257
### テストに成功する画像
https://gyazo.com/ecc4e84f697722af885f02dde657e29b

nana0817 committed:
deviseのインストール

nana0817 committed:
Userモデルの作成

nana0817 committed:
usersテーブルの作成

nana0817 committed:
新しく追加したカラムをストロングパラメーターに設定

nana0817 committed:
ログインの有無でヘッダーの表示が変わるように変更

ユーザーがログインしていれば、ユーザーのnicknameと「ログアウト」の文字、
ユーザーがログインしていなければ、「ログイン」と「新規登録」の文字が表示されるよう変更。
また、ログアウト・ログイン・新規登録に関しては、パスを設定。

nana0817 committed:
_error_messages.html.erb　インデントの調整

nana0817 committed:
新規登録のnew.html.erbの編集

・form_withに適切な値を記述
・エラー発生時にメッセージが表示されるよう変更
・hogeとなっていたname属性を適切な値に変更
・可読性を上げるためにコメントアウトを記述

nana0817 committed:
ログアウト機能の実装

nana0817 committed:
ログインのnew.html.erbの編集

・form_withの中身をインデント
・可読性を上げるためにコメントアウトを記述
・name属性の値を適切なものへ変更

nana0817 committed:
Userモデルのバリデーションを設定

nana0817 committed:
長くて読みづらいので改行

nana0817 committed:
ログインのnew.html.erb　修正

modelに@user、
urlにdevise/session#createのパスを指定

nana0817 committed:
エラーハンドリングのため、usersコントローラーを作成

nana0817 committed:
カラム名の変更

全角漢字フルネームと全角カナフルネームのカラム名が長くて冗長だったため、変更

nana0817 committed:
アンダーバーを２個つけてしまっていたので修正

nana0817 committed:
デバッグのためにpry-railsを導入

nana0817 committed:
カラムの変更に伴った修正

nana0817 committed:
rails devise:controllers usersで作成したコントローラーを削除

新規登録・ログインに失敗した際に入力欄に内容が残っている状態にすることは通常のdeviseの機能として備わっているため、新たにコントローラーを作成した、renderメソッドで条件分岐を組む必要はない。

nana0817 committed:
Userモデルのテストコード実装のための準備

以下のことを実行
・テストに必要なGem(Rspec, FactoryBot, Faker)の導入
・rails g rspec:installでRspecをインストール
・.rspecファイルに設定を追加
・テストファイルを作成(spec/models/user_spec.rb)
・app/models/user.rbのnicknameカラムに文字数制限のバリデーション追加
・新規登録のビューファイルのnicknameのフォームに設けている文字数制限の記述を削除
・spec/factories/users.rbファイルにFakerとGimeiを使って、ランダムな値を生成する記述を追加
・spec/rails_helper.rbファイルにエラーメッセージを英語に設定する記述を追加
・テストファイルにテストコードの骨組みを作成、exampleを記述
・カラム名を変更

nana0817 committed:
バリデーションを追加

以下のバリデーションを新たに追加
last_name, first_name
↓
正規表現を用いて、全角(漢字・ひらがな・カタカナ)のみ使えるようにした

last_name_kana, first_name_kana
↓
正規表現を用いて、全角(カタカナ)のみ使えるようにした

nana0817 committed:
nicknameを自動生成するFakerのモジュールを変更

モジュールをNameからInternetに変更。

nana0817 committed:
Gem webdriversの削除

テストコードの実行に支障が出るので削除

nana0817 committed:
カラム名の修正

family_name → last_name
family_name_kana → last_name_kana
と修正するのを忘れていた。

nana0817 committed:
バリデーションの追加・修正

パスワードのバリデーションを正規表現を用いて追加

また、フルネームのバリデーションに不備があったので、それを修正

nana0817 committed:
Userモデルの単体テストコード実装

nana0817 committed:
Rubocopの拡張機能導入

nana0817 committed:
Rubocopによるコードの自動修正

ShotaOtake commented on spec/factories/users.rb:
```diff
@@ -0,0 +1,14 @@
+  factory :user do
+    nickname              { Faker::Internet.username }
+    email                 { Faker::Internet.email }
+    password              { Faker::Internet.password(min_length: 6) }
```
◎passwordカラムを修正しましょう。
現状、必ずしも半角英数混合になるかどうかが不明となる記述です。
(Fakerのみの場合、英字のみのパスワードが生成されてしまうことがあるためです。)
確実に半角英数混合の値を生成するため、『+』の記号を使用し、文字列を組み合わせることで期待する値を生成する内容に修正していきましょう。

ShotaOtake commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,130 @@
+        expect(@user.errors.full_messages).to include('Password is too long (maximum is 128 characters)')
+      end
+
+      it 'passwordは半角英数字混合でなければ登録できない' do
```
◎異常系テストを追加しましょう。

■理由
下記要件を満たして頂くためです。
・英字のみのパスワードでは登録できない
・数字のみのパスワードでは登録できない
・全角文字を含むパスワードでは登録できない

※パスワードにつきましては上記それぞれの項目に分けてexampleの生成が必要です。

ShotaOtake reviewed (commented):
レビュー依頼を頂きありがとうございます！
追加で修正依頼のコメントをさせて頂きましたので、ご確認の上修正をお願い致します。

完成まであと一歩です！
引き続き実装頑張っていきましょう！

```
※テストコードにて修正箇所のございます際は、お手数ですが、最新の実行結果の画像も改めて添付頂きますようお願い致します。
```

nana0817 commented:
コードレビューありがとうございます！
早速修正に取り掛かります！

nana0817 committed:
FactoryBotで自動生成するpasswordの記述を変更

必ず半角英数字混合になるか不明な記述であったため、変更(英字のみ、数字のみになる可能性があった)。

nana0817 committed:
テストコード修正

元々のexampleを以下の３つに細分化した。
・passwordは英字のみでは登録できない
・passwordは数字のみでは登録できない
・passwordは全角文字では登録できない

加えてそれぞれを検証するテストコードを追加した。

nana0817 commented:
ご指摘いただいた点は全て修正いたしました。

新しいテストコード実行結果の画像を添付します。
https://i.gyazo.com/860ce16047b567e31096b9676a6d8de2.png

AsukaToyoshima reviewed (commented):
レビューのご依頼ありがとうございます。
LGTMとさせていただきます。お疲れ様でした！
引き続き、実装頑張っていきましょう！　応援しております！
